tests/server/metrics-unit.js

```diff
@@ -0,0 +1,231 @@
+    var test = setUp(function (headerName) {
+      if (headerName.toLowerCase() === 'content-type') {
+        return 'text/plain';
+      }
+
+      return 'foo';
+    });
+    test.mocks.request.body = 'bar';
+    test.metrics.process(test.mocks.request, test.mocks.response);
+    test.callbacks.nextTick();
+
+    assert.equal(test.mocks.logger.error.callCount, 1);
+    assert.lengthOf(test.mocks.logger.error.getCall(0).args, 1);
+    assert.instanceOf(test.mocks.logger.error.getCall(0).args[0], Error);
+
+    assert.equal(test.mocks.metricsCollector.write.callCount, 0);
+    assert.equal(test.mocks.statsdCollector.write.callCount, 0);
+    assert.equal(test.mocks.gaCollector.write.callCount, 0);
+
+    tearDown();
+  };
+
+  suite['Content-Type is text/plain, data is valid JSON, user is sampled'] = function () {
+    var test = setUp(function (headerName) {
+      if (headerName.toLowerCase() === 'content-type') {
+        return 'text/plain';
+      }
+
+      return 'wibble';
+    });
+    test.mocks.request.body = '{"foo":"bar","isSampledUser":true}';
+    test.metrics.process(test.mocks.request, test.mocks.response);
+    test.callbacks.nextTick();
+
+    assert.equal(test.mocks.logger.error.callCount, 0);
+
+    assert.equal(test.mocks.metricsCollector.write.callCount, 1);
+    var data = test.mocks.metricsCollector.write.getCall(0).args[0];
+    assert.lengthOf(Object.keys(data), 3);
+    assert.equal(data.agent, 'wibble');
+    assert.equal(data.foo, 'bar');
+    assert.equal(data.isSampledUser, true);
+
+    assert.equal(test.mocks.statsdCollector.write.callCount, 1);
+    assert.equal(test.mocks.statsdCollector.write.getCall(0).args[0], data);
+
+    assert.equal(test.mocks.gaCollector.write.callCount, 1);
+    assert.equal(test.mocks.gaCollector.write.getCall(0).args[0], data);
+
+    tearDown();
+  };
+
+  suite['Content-Type is text/plain;charset=UTF-8, data is valid JSON, user is sampled'] = function () {
+    var test = setUp(function (headerName) {
+      if (headerName.toLowerCase() === 'content-type') {
+        return 'text/plain;charset=UTF-8';
+      }
+
+      return 'foo';
+    });
+    test.mocks.request.body = '{"isSampledUser":true}';
+    test.metrics.process(test.mocks.request, test.mocks.response);
+    test.callbacks.nextTick();
+
+    assert.equal(test.mocks.logger.error.callCount, 0);
+
+    assert.equal(test.mocks.metricsCollector.write.callCount, 1);
+    var data = test.mocks.metricsCollector.write.getCall(0).args[0];
+    assert.lengthOf(Object.keys(data), 2);
+    assert.equal(data.agent, 'foo');
+    assert.equal(data.isSampledUser, true);
+
+    assert.equal(test.mocks.statsdCollector.write.callCount, 1);
+    assert.equal(test.mocks.statsdCollector.write.getCall(0).args[0], data);
+
+    assert.equal(test.mocks.gaCollector.write.callCount, 1);
+    assert.equal(test.mocks.gaCollector.write.getCall(0).args[0], data);
+
+    tearDown();
+  };
+
+  registerSuite(suite);
+
+  function setUp (requestGet) {
```

Comment:
@shane-tomlinson: I don't care if you use a named function before it is defined.

Ref: [**eslint-config-fxa** /index.js:40](https://github.com/pdehaan/eslint-config-fxa/blob/b055709b3142945e0863f1f1efa93e1b1acb87cb/index.js#L40):
```js
'no-use-before-define': [2, 'nofunc'],
```

> `no-use-before-define`: "It takes an optional option as the second parameter which can be `"nofunc"` to allow named function definitions to be used before the location where they are defined."
> &mdash; via http://eslint.org/docs/rules/no-use-before-define

Let me know if you want that changed and I can override to be more strict in the .eslintrc file or more globally in **eslint-config-fxa** /client.js.